### ISSUE TITLE
Fresh install of microceph, database not ready

### ISSUE THREAD

**Issue by fzhan** [canonical/microceph#370]
# Issue report

## What version of MicroCeph are you using ?
`
name:      microceph
summary:   Simple clustered Ceph deployment
publisher: Canonical✓
store-url: https://snapcraft.io/microceph
contact:   https://matrix.to/#/#ubuntu-ceph:matrix.org
license:   AGPL-3.0
description: |
  MicroCeph is the easiest way to get up and running with Ceph. It is focused on providing a modern
  deployment and management experience to Ceph administrators and storage software developers.

  The below commands will set you up with a testing environment on a single machine using
  file-backed OSDs - you'll need about 15 GiB of available space on your root drive:

      sudo snap install microceph
      sudo snap refresh --hold microceph
      sudo microceph cluster bootstrap
      sudo microceph disk add loop,4G,3
      sudo ceph status

  You're done!

  You can remove everything cleanly with:

      sudo snap remove microceph

  To learn more about MicroCeph see the documentation:

  https://canonical-microceph.readthedocs-hosted.com
commands:
  - microceph.ceph
  - microceph
  - microceph.rados
  - microceph.radosgw-admin
  - microceph.rbd
services:
  microceph.daemon: simple, enabled, active
  microceph.mds:    simple, disabled, inactive
  microceph.mgr:    simple, disabled, inactive
  microceph.mon:    simple, disabled, inactive
  microceph.osd:    simple, disabled, inactive
  microceph.rgw:    simple, disabled, inactive
snap-id:      ct1DgtIGBaljdOwomQGMLr8EJcL5pOP1
tracking:     reef/stable
refresh-date: today at 13:22 AEST
hold:         forever
channels:
  reef/stable:      18.2.0+snap71f71782c5     2024-05-28  (981) 87MB -
  reef/candidate:   18.2.0+snapcba31e8c75     2024-06-05  (999) 87MB -
  reef/beta:        18.2.0+snap38155f2c4e     2024-06-07 (1015) 87MB -
  reef/edge:        18.2.0+snap556b907075     2024-06-12 (1026) 86MB -
  latest/stable:    18.2.0+snap71f71782c5     2024-05-28  (981) 87MB -
  latest/candidate: 18.2.0+snapcba31e8c75     2024-06-05  (999) 87MB -
  latest/beta:      ↑
  latest/edge:      18.2.0+snap556b907075     2024-06-12 (1026) 86MB -
  squid/stable:     –
  squid/candidate:  –
  squid/beta:       –
  squid/edge:       19.2.0~git+snap44f97f5158 2024-06-06 (1011) 89MB -
  quincy/stable:    0+git.4a608fc             2024-01-10  (793) 96MB -
  quincy/candidate: 0+git.4a608fc             2023-11-30  (793) 96MB -
  quincy/beta:      ↑
  quincy/edge:      0+git.287ee68             2023-12-05  (807) 96MB -
installed:          18.2.0+snap71f71782c5                 (981) 87MB held
`

## What are the steps to reproduce this issue ?
1. sudo snap install microceph
2. sudo snap refresh --hold microceph

## What happens (observed behaviour) ?
journalctl -f -u snap.microceph.daemon.service
"start: database not ready, waiting..."

## What were you expecting to happen ?
microceph runs without error

**Comment by UtkarshBhatthere** [comment]
Can you provide a brief description of what faults you are seeing in MicroCeph operations ?

**Comment by fzhan** [comment]
@UtkarshBhatthere the daemon could not start, so I checked the log and it just kept saying 'database not ready'. Every ceph command just reports the service is not ready.

It's a completely fresh install of anUbuntu 22.04 and microceph.

**Comment by UtkarshBhatthere** [comment]
umm, the database would only be ready post bootstrap. This error should go away once you `sudo microceph cluster bootstrap` ?

**Comment by fzhan** [comment]
@UtkarshBhatthere the tutorial is not clear for this action. https://canonical-microceph.readthedocs-hosted.com/en/reef-stable/tutorial/multi-node/

However, after I've ran `sudo microceph cluster bootstrap` on the new node, and `microceph cluster add node2`, `microceph cluster list` now listed the node2 as pending, with status `unreachable`. Whilst on node2, microceph cluster join TOKEN reported:
Error: A remote with name "node2." already exists.

How to replicate:
1. One of the 5-node cluster was removed, disk had to be replaced and system reinstalled
2. That node was never successfully removed from the cluster
3. Tried to add the node back in but MicroCeph refuses to talk to the new one

How can I:
1. Remove the old server from the cluster list? It doesn't allow me since that old server had mon service on 
2. Add the new server back in

**Comment by fzhan** [comment]
cluster just doesn't know how to proceed:
`
root@node1:/home/xxx# microceph cluster add node2
Error: Failed to create "internal_token_records" entry: UNIQUE constraint failed: internal_token_records.name
root@node1:/home/xxx# microceph cluster remove node2
Error: No remote exists with the given name "node2"
root@node1:/home/xxx#
`

**Comment by UtkarshBhatthere** [comment]
if you know your way around SQL, It should be possible to forcefully remove the bad actor from the internal member table.

An example of how you can easily run SQL queries against the internal table is this.
```
$ sudo microceph cluster sql "select * from internal_cluster_members"
```

**Comment by fzhan** [comment]
@UtkarshBhatthere thank you! I feel like I'm on the next level!

Is there list of SQL tables somewhere in the doc or should I just explore myself :) Because the node server also reported error when try to join cluster 
```
"Error: A remote with name "node2" already exists
```

**Comment by fzhan** [comment]
It's working after I removed the record in the master node, and reinstall microceph in the newly join node.

1. Remove the existing token / record in the sql table, from "internal_cluster_members" and "internal_token_records"
2. Remove microceph from the new node "snap remove microceph"
3. Install microceph again in the new node
4. Back to master node and follow "add node" instruction

**Comment by UtkarshBhatthere** [comment]
happy that your hacking adventures went great @fzhan!

**Comment by fzhan** [comment]
@UtkarshBhatthere I've ran into the same issue, but this time, "Database is waiting for an upgrade: 3 cluster members have not yet received the update". How can I manually sync the database?

**Comment by sabaini** [comment]
@fzhan this means that your cluster nodes are not at the same snap revision. The MicroCeph configuration database requires that all node members be at the same version to prevent inconsistencies. 
You can remedy this by refreshing the snaps on all nodes to the same version via `snap refresh microceph` in sequence. As the refresh will cause a restart of the Ceph services, ensure that each node finishes refresh before starting refreshing the next one

**Comment by fzhan** [comment]
Hi @sabaini 
You are absolutely right! Thought I had all of them put on hold but apparently one slipped (the one was causing issue joining at the beginning of this thread.

**Comment by fzhan** [comment]
T_T

Stuck on 'Copy snap data', done upgrading one of them, then
```
 snap refresh microceph --channel=reef/stable
2024-08-14T09:34:32+10:00 INFO Waiting for "snap.microceph.mon.service" to stop.
2024-08-14T09:34:33+10:00 INFO Waiting for "snap.microceph.osd.service" to stop.
2024-08-14T09:34:37+10:00 INFO Waiting for "snap.microceph.mds.service" to stop.
Copy snap "microceph" data 
```

What should I do?

**Comment by sabaini** [comment]
@fzhan the err="schema check gracefully aborted" message is the database check to see if all nodes are on the same revision -- it only means that not all nodes are on the same db version and is otherwise benign. 

As for the snap upgrade, hard to tell, will need some troubleshooting -- is there anything interesting in the logs regarding the snap operation? what does `snap changes` say? Might need to doublecheck config issues and disk capacity
